w1r3/python/w1r3.py

```diff
@@ -0,0 +1,377 @@
+# limitations under the License.
+
+"""The Python version of the W1R3 benchmark."""
+
+import argparse
+import random
+import resource
+import sys
+import time
+import uuid
+
+import google.cloud.storage as storage
+import google.protobuf
+import grpc
+import opentelemetry.metrics as metrics
+import opentelemetry.sdk.metrics as sdkmetrics
+import opentelemetry.sdk.metrics.view as view
+import opentelemetry.sdk.trace as sdktrace
+import opentelemetry.trace as trace
+import requests
+from opentelemetry.exporter.cloud_monitoring import CloudMonitoringMetricsExporter
+from opentelemetry.exporter.cloud_trace import CloudTraceSpanExporter
+from opentelemetry.instrumentation.requests import RequestsInstrumentor
+from opentelemetry.resourcedetector.gcp_resource_detector import (
+    GoogleCloudResourceDetector,
+)
+from opentelemetry.sdk.metrics.export import PeriodicExportingMetricReader
+from opentelemetry.sdk.resources import Resource
+from opentelemetry.sdk.trace.export import BatchSpanProcessor
+
+# You can optionally pass a custom TracerProvider to instrument().
+response = requests.get(url="https://www.example.org/")
+
+
+def main():
+    args = parse_args()
+    instance = str(uuid.uuid4())
+    print("## Starting continuous GCS Python SDK benchmark")
+    print(f"# object-sizes: {args.object_size}")
+    print(f"# transports: {args.transport}")
+    print(f"# project-id: {args.project_id}")
+    print(f"# bucket: {args.bucket}")
+    print(f"# deployment: {args.deployment}")
+    print(f"# instance: {instance}")
+    print(f"# Benchmark Version: {_SSB_W1R3_VERSION}")
+    print(f"# Python Version: {sys.version}")
+    print(f"# Python SDK Version: {storage.__version__}")
+    print(f"# gRPC Version: {grpc.__version__}")
+    print(f"# Protobuf Version: {google.protobuf.__version__}")
+    print(f"# HTTP Client Version: {requests.__version__}")
+
+    tracer = configure_trace_exporter(args)
+    (meter_provider, meter) = configure_metrics_exporter(instance, args)
+
+    # Instrument Python's `requests` library to generate spans.
+    RequestsInstrumentor().instrument()
+
+    clients = make_clients(args)
+    prng = random.Random()
+    prng.seed()
+    random_data = bytes([prng.randint(0, 255) for _ in range(0, max(args.object_size))])
+
+    worker(clients, instance, prng, random_data, tracer, meter, args)
+
+
+def worker(
+    clients,
+    instance,
+    prng,
+    random_data,
+    tracer: sdktrace.Tracer,
+    meter,
+    args: argparse.Namespace,
+):
+    latency = meter.create_histogram(
+        _LATENCY_HISTOGRAM_NAME,
+        unit=_LATENCY_HISTOGRAM_UNIT,
+        description=_LATENCY_HISTOGRAM_DESCRIPTION,
+    )
+    cpu = meter.create_histogram(
+        _CPU_HISTOGRAM_NAME,
+        unit=_CPU_HISTOGRAM_UNIT,
+        description=_CPU_HISTOGRAM_DESCRIPTION,
+    )
+
+    for i in range(args.iterations):
+        (transport_name, client) = prng.choice(clients)
+        object_size = prng.choice(args.object_size)
```

Comment:
Just wanted to confirm if my understanding is right… so we’re not randomly choosing from a range, but instead from a list of object sizes, e.g  100KB, 2MiB, 100MB

Reply:
Yes. We want to upload the object, a range won't be as useful.  We have lots of data showing nothing changes much if you do range reads of size X vs. reading an object of size X.